Return-Path: <xorg-devel-bounces+chase.douglas=canonical.com@lists.x.org>
X-Original-To: cndougla@mail.canonical.com
Delivered-To: cndougla@mail.canonical.com
Received: from fiordland.canonical.com (fiordland.canonical.com [91.189.94.145])
	by grenadilla.canonical.com (Postfix) with ESMTP id 06AB914721B9
	for <cndougla@mail.canonical.com>; Fri, 10 Feb 2012 17:11:13 +0000 (UTC)
Received: from cluster-e.mailcontrol.com (cluster-e.mailcontrol.com [85.115.58.190])
	by fiordland.canonical.com (Postfix) with ESMTP id EEEDFA183BE
	for <chase.douglas@cleanmail.canonical.com>; Fri, 10 Feb 2012 17:11:12 +0000 (UTC)
Received: from arctowski.canonical.com (arctowski.canonical.com [91.189.94.158])
	by rly08e.srv.mailcontrol.com (MailControl) with ESMTP id q1AHBB8v013234
	for <chase.douglas@cleanmail.canonical.com>; Fri, 10 Feb 2012 17:11:11 GMT
Received: from fiordland.canonical.com ([91.189.94.145])
	by arctowski.canonical.com with esmtp (Exim 4.71)
	(envelope-from <xorg-devel-bounces+chase.douglas=canonical.com@lists.x.org>)
	id 1Rvu07-0008BT-Bu
	for chase.douglas@cleanmail.canonical.com; Fri, 10 Feb 2012 17:11:11 +0000
Received: from gabe.freedesktop.org (gabe.freedesktop.org [131.252.210.177])
	by fiordland.canonical.com (Postfix) with ESMTP id 007A4A183C9
	for <chase.douglas@canonical.com>; Fri, 10 Feb 2012 17:11:11 +0000 (UTC)
Received: from gabe.freedesktop.org (localhost [127.0.0.1])
	by gabe.freedesktop.org (Postfix) with ESMTP id 7E3AAA0D8B
	for <chase.douglas@canonical.com>; Fri, 10 Feb 2012 09:11:10 -0800 (PST)
X-Original-To: xorg-devel@lists.freedesktop.org
Delivered-To: xorg-devel@lists.freedesktop.org
Received: from mail.clearchain.com (leo.clearchain.com [199.73.29.74])
	by gabe.freedesktop.org (Postfix) with ESMTP id 75D04A0DA7
	for <xorg-devel@lists.freedesktop.org>;
	Fri, 10 Feb 2012 09:10:29 -0800 (PST)
Received: from leo.clearchain.com (localhost [127.0.0.1])
	by mail.clearchain.com (8.14.5/8.14.5) with ESMTP id q1AHARrY005234
	(version=TLSv1/SSLv3 cipher=DHE-RSA-AES256-SHA bits=256 verify=NO)
	for <xorg-devel@lists.freedesktop.org>;
	Sat, 11 Feb 2012 03:40:27 +1030 (CST)
	(envelope-from peter.hutterer@who-t.net)
Received: (from whot@localhost)
	by leo.clearchain.com (8.14.5/8.14.5/Submit) id q1AHARTs005232
	for xorg-devel@lists.freedesktop.org;
	Sat, 11 Feb 2012 03:40:27 +1030 (CST)
	(envelope-from peter.hutterer@who-t.net)
X-Authentication-Warning: leo.clearchain.com: whot set sender to
	peter.hutterer@who-t.net using -f
Date: Sat, 11 Feb 2012 03:12:28 +1000
From: Peter Hutterer <peter.hutterer@who-t.net>
To: "X.Org Devel List" <xorg-devel@lists.freedesktop.org>
Subject: [PATCH] dix: reset last.scroll when resetting the valuator (#45611)
Message-ID: <20120210171228.GA5823@yabbi.redhat.com>
MIME-Version: 1.0
Content-Disposition: inline
User-Agent: Mutt/1.5.21 (2010-09-15)
X-Greylist: Sender is SPF-compliant, not delayed by milter-greylist-4.0.1
	(mail.clearchain.com [127.0.0.1]);
	Sat, 11 Feb 2012 03:40:28 +1030 (CST)
X-BeenThere: xorg-devel@lists.x.org
X-Mailman-Version: 2.1.11
Precedence: list
List-Id: "X.Org development list" <xorg-devel.lists.x.org>
List-Unsubscribe: <http://lists.x.org/mailman/options/xorg-devel>,
	<mailto:xorg-devel-request@lists.x.org?subject=unsubscribe>
List-Archive: <http://lists.x.org/archives/xorg-devel>
List-Post: <mailto:xorg-devel@lists.x.org>
List-Help: <mailto:xorg-devel-request@lists.x.org?subject=help>
List-Subscribe: <http://lists.x.org/mailman/listinfo/xorg-devel>,
	<mailto:xorg-devel-request@lists.x.org?subject=subscribe>
Content-Type: text/plain; charset="us-ascii"
Content-Transfer-Encoding: 7bit
Sender: xorg-devel-bounces+chase.douglas=canonical.com@lists.x.org
Errors-To: xorg-devel-bounces+chase.douglas=canonical.com@lists.x.org
X-Mailcontrol-Inbound: uq3drnD2P+ps5SfEb0fvr78+NoP1DHBZwGqKpaXB2eTgNv8D6KLIxb8+NoP1DHBZ8VSaBg0k0xw=
X-Spam-Score: -0.5
X-Scanned-By: MailControl 7.6.6 (www.mailcontrol.com) on 10.69.0.118

last.scroll remained on the last-submitted scrolling value but last.valuator
was changed whenever the slave device changed. The first scrolling delta
after a switch was then calculated as (last.scroll - new abs value), causing
erroneous scrolling events.

Test case:
- synaptics with a scrolling method enabled, other device with 3+ axes (e.g.
  wacom)
- scroll on touchpad
- use other device
- scroll on touchpad

The second scroll caused erroneous button press/release events.

X.Org Bug 45611 <http://bugs.freedesktop.org/show_bug.cgi?id=45611>

Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>
Tested-by: Timo Aaltonen <timo.aaltonen@canonical.com>
---
 dix/getevents.c |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/dix/getevents.c b/dix/getevents.c
index 7678aa1..6ea4ba0 100644
--- a/dix/getevents.c
+++ b/dix/getevents.c
@@ -360,6 +360,7 @@ updateSlaveDeviceCoords(DeviceIntPtr master, DeviceIntPtr pDev)
             if (i >= lastSlave->valuator->numAxes)
             {
                 pDev->last.valuators[i] = 0;
+                valuator_mask_set_double(pDev->last.scroll, i, 0);
             }
             else
             {
@@ -367,6 +368,7 @@ updateSlaveDeviceCoords(DeviceIntPtr master, DeviceIntPtr pDev)
                 val = rescaleValuatorAxis(val, lastSlave->valuator->axes + i,
                                           pDev->valuator->axes + i, 0, 0);
                 pDev->last.valuators[i] = val;
+                valuator_mask_set_double(pDev->last.scroll, i, val);
             }
         }
     }
-- 
1.7.7.5
_______________________________________________
xorg-devel@lists.x.org: X.Org development
Archives: http://lists.x.org/archives/xorg-devel
Info: http://lists.x.org/mailman/listinfo/xorg-devel

